wm.c

```diff
@@ -42,41 +56,77 @@ int main() {
 	FARPROC shellProc = GetProcAddress(wmDll, "ShellProc");
 
 	if (shellProc == NULL) { 
-		reportWin32Error(L"GetProcAddress");
+		reportWin32Error(L"GetProcAddress failed for shell even callback");
 		goto cleanup; 
 	}
+	
+	HotKeyProcType HotKeyProc = (HotKeyProcType)GetProcAddress(wmDll, "HotkeyProc");
 
+	if (HotKeyProc == NULL) { 
+		reportWin32Error(L"GetProcAddress failed for shell even callback");
+		goto cleanup; 
+	}
+	
 	windowEvent = CreateEventW(NULL, FALSE, FALSE, L"LightWMWindowEvent");
 
 	if (windowEvent == NULL) {
 		reportWin32Error(L"CreateEvent");
 		goto cleanup;
 	}
 
-	hookHandle = SetWindowsHookExW(WH_SHELL, (HOOKPROC)shellProc, wmDll, 0);
+	hookShellProcHandle = SetWindowsHookExW(WH_SHELL, (HOOKPROC)shellProc, wmDll, 0);
 
-	if (hookHandle == NULL) {
-		reportWin32Error(L"SetWindowsHookExW");
+	if (hookShellProcHandle == NULL) {
+		reportWin32Error(L"SetWindowsHookExW failed for shell hook");
 		goto cleanup;
 	}
-
+	
 	signal(SIGINT, ctrlc);
-
+ 
+	//Load the configuration
+	if(LoadConfigFile(NULL) != ERROR_SUCCESS) 
+	{ 
+		reportWin32Error(L"Load config file");
+		goto cleanup; 
+	}
+	
+	if(!InitializeKeyboardConfig(GetConfigItems())) 
+	{ 
+		reportWin32Error(L"Setup keyboard config"); 
+		goto cleanup; 
+	}
+ 
+	// Handle a message loop
 	tileWindows();
-
-	for (;;) {
-		if (WaitForSingleObject(windowEvent, INFINITE) == WAIT_FAILED) {
+	MSG msg; 
+	while (GetMessage(&msg, NULL, 0, 0) != 0) {
+		if(msg.message == WM_HOTKEY) { 
+			//Because Win32 doesn't support hook callbacks with RegisterHotkey lets make our own callback. 
+			assert(HotKeyProc != NULL); 
+			
+			LRESULT ret = HotKeyProc(0, msg.wParam, msg.lParam);
+			if(ret != ERROR_SUCCESS) { 
+				DEBUG_PRINT("HotKey was unhandled! Ret: %i", ret); 
+			}
+			
+			TranslateMessage(&msg); 
+			DispatchMessageW(&msg); 
+			continue;
+		} else if (WaitForSingleObject(windowEvent, INFINITE) == WAIT_FAILED) {
```

Comment:
This will be blocking the message loop until the next time the event is signaled, and also the getmessage will be blocking the event from being checked.

A solution for this can be to create an new thread for handling the hot keys.

Reply:
Since RegisterHotkey uses the message queue when a user presses a hotkey it puts a message into the message queue, hence why using a hook is impossible with the RegisterHotkey functionality of the Windows API. GetMessage automatically gets fired for this thread when there is a message available, and currently the only two messages that do get fired are a WM_HOTKEY message and the windowEvent message, its a simple check if the msg is of type hotkey it calls the custom callback hook in the dll with the id of the registered hotkey, then goes back to waiting for a message. 

Like I commented earlier if we resort to forgoing the RegisterHotkey and message queue functionality of the Windows API custom hotkey registration and handling functionality will have to be implemented leading to much more complexity.

Reply:
The windowEvent is not sent as a message, it is a handle to an event, hence, getmessage will not return if the event is signaled and currently this logic will delay the hotkey handling or delay the window event handling (since also get message waits and also waitforsingleobject waits).

The solution for this problem (since both are currently blocking each other currently), is to do one of them in a separate thread.

Reply:
BTW, no need to call DispatchMessage, there is no window procedure. Also, no need for TranslateMessage.

Reply:
A quick question, what was the reason to use an event instead of posting a message ?

Reply:
Originally I used an event since it is simple to work with, but actually posting a thread message is a good idea! (Could keep the logic simple rather than adding a new thread) - only thing is that it would require passing getting the thread id of the window manager from the DLL - but that could easily done by adding another function to the DLL that sets this as a global variable.